### PR TITLE
feat(contracts-rfq): arbitrary calls without additional native value [SLT-233]

### PR DESCRIPTION
**Pull request by ChiTimesChi**
**Description**
This is the first of the two PRs for the "Arbitrary Call" (final name pending) feature. The first PR adds the option to invoke a special function hook once the tokens have been transferred to the recipient:

https://github.com/synapsecns/sanguine/blob/ad7044902e50154faac6022df7d6eae756c699db/packages/contracts-rfq/contracts/interfaces/IFastBridgeRecipient.sol#L4-L13

The origin sender is given an option to provide `callParams` when initiating a bridge on the source chain.
- If nothing is passed, the hook function is **not** called, and this functions exactly as the existing (legacy) workflow.
- If `callParams` is non-empty, `destRecipient.fastBridgeTransferReceived()` function is called after the tokens are transferred to the recipient
  - If token is a native gas token (e.g. ETH), it is supplied as `msg.value` for the hook function call to avoid calling the recipient twice.

The transactions with a hook function call will revert in one of these cases (this does not affect the legacy transactions):
- Recipient is EOA
- Recipient does not implement `fastBridgeTransferReceived` function, or doesn't return the magic value (the hook function selector) at the end of the call
- Recipient reverts within the `fastBridgeTransferReceived` call

The reverts do not put Relayer or Recipient funds at risk, in fact the transaction is only completed if the tokens are transferred to the recipient **AND** the hook call is successful (as reported by the recipient itself using the magic return value). The relayers are expected to simulate the transaction prior its submission to minimise the chances of the on-chain reverts.

## Summary by CodeRabbit

- **New Features**
	- Enhanced `FastBridgeV2` contract to support additional call parameters during bridge transactions.
	- Introduced a new interface `IFastBridgeRecipient` for handling token transfer receipts.
	- Added a new contract for testing various scenarios related to token and ETH transfers with arbitrary calls.
	- Implemented new functions for setting exclusivity parameters in tests.
	- Expanded functionality in testing contracts to validate behavior with maximum call parameter lengths and negative exclusivity periods.

- **Bug Fixes**
	- Improved error handling with new error declarations for call parameter validation.

- **Tests**
	- Expanded testing framework to validate the behavior of the bridge with new call parameters and recipient scenarios, including edge cases for length validation and exclusivity conditions.

### CODE REVIEW

File: packages/contracts-rfq/contracts/FastBridgeV2.sol [hunk 1]
```diff
@@ -132,6 +142,7 @@ contract FastBridgeV2 is Admin, IFastBridgeV2, IFastBridgeV2Errors {
         if (params.sender == address(0) || params.to == address(0)) revert ZeroAddress();
         if (params.originToken == address(0) || params.destToken == address(0)) revert ZeroAddress();
         if (params.deadline < block.timestamp + MIN_DEADLINE_PERIOD) revert DeadlineTooShort();
+        if (paramsV2.callParams.length > MAX_CALL_PARAMS_LENGTH) revert CallParamsLengthAboveMax();
         int256 exclusivityEndTime = int256(block.timestamp) + paramsV2.quoteExclusivitySeconds;
         // exclusivityEndTime must be in range (0 .. params.deadline]
         if (exclusivityEndTime <= 0 || exclusivityEndTime > int256(params.deadline)) {
```
[the remark this answers is not in the document]
Comment (trajan0x) [reply packages/contracts-rfq/contracts/FastBridgeV2.sol:148]: Solidity ^0.8 prevents overflows by trigering hard revert.

File: packages/contracts-rfq/contracts/FastBridgeV2.sol [hunk 2]
```diff
@@ -327,6 +351,31 @@
...
+        bytes memory returnData = Address.functionCallWithValue({target: recipient, data: hookData, value: msgValue});
+        // Explicit revert if no return data at all
+        if (returnData.length == 0) revert RecipientNoReturnValue();
+        // Check that exactly a single return value was returned
+        if (returnData.length != 32) revert RecipientIncorrectReturnValue();
+        // Return value should be abi-encoded hook function selector
+        if (bytes32(returnData) != bytes32(IFastBridgeRecipient.fastBridgeTransferReceived.selector)) {
+            revert RecipientIncorrectReturnValue();
+        }
+    }
```
[the remark this answers is not in the document]
Comment (trajan0x) [reply packages/contracts-rfq/contracts/FastBridgeV2.sol:380]: I'm also a bit surprised this isn't non-reenrant. Can you explain why?

**Review by trajan0x — CHANGES_REQUESTED** [review]

File: packages/contracts-rfq/contracts/FastBridgeV2.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-rfq/contracts/FastBridgeV2.sol:380]: @trajan0x this PR doesn't add any additional risk vectors: legacy ETH transfers are also done through the recipient call. Therefore no additional reenrancy protection was introduced here. I suggest to leave it out of this PR as being a potentially more of a global thing we might or might not want to deal with.

**Review by parodime — APPROVED** [review]
